raft/src/main/java/org/apache/kafka/raft/DynamicVoter.java

```diff
@@ -0,0 +1,188 @@
+                        "bracket was found.");
+            }
+            host = input.substring(1, endBracketIndex);
+            input = input.substring(endBracketIndex + 1);
+        } else {
+            int endColonIndex = input.indexOf(":");
+            if (endColonIndex < 0) {
+                throw new IllegalArgumentException("No colon following hostname could be found.");
+            }
+            host = input.substring(0, endColonIndex);
+            input = input.substring(endColonIndex);
+        }
+        if (!input.startsWith(":")) {
+            throw new IllegalArgumentException("Port section must start with a colon.");
+        }
+        input = input.substring(1);
+        int endColonIndex = input.indexOf(":");
+        if (endColonIndex < 0) {
+            throw new IllegalArgumentException("No colon following port could be found.");
+        }
+        String portString = input.substring(0, endColonIndex);
+        int port;
+        try {
+            port = Integer.parseInt(portString);
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("Failed to parse port in dynamic voter string.", e);
+        }
+        if (port < 0 || port > 65535) {
+            throw new IllegalArgumentException("Invalid port " + port + " in dynamic voter string.");
+        }
+        String directoryIdString = input.substring(endColonIndex + 1);
+        Uuid directoryId;
+        try {
+            directoryId = Uuid.fromString(directoryIdString);
+        } catch (IllegalArgumentException e) {
+            throw new IllegalArgumentException("Failed to parse directory ID in dynamic voter string.", e);
+        }
+        return new DynamicVoter(directoryId, nodeId, host, port);
+    }
+
+    /**
+     * Create a new KIP-853 voter.
+     *
+     * @param directoryId   The directory ID.
+     * @param nodeId        The voter ID.
+     * @param host          The voter hostname or IP address.
+     * @param port          The voter port.
+     */
+    public DynamicVoter(
+        Uuid directoryId,
+        int nodeId,
+        String host,
+        int port
+    ) {
+        this.directoryId = directoryId;
+        this.nodeId = nodeId;
+        this.host = host;
+        this.port = port;
+    }
+
+    public Uuid directoryId() {
+        return directoryId;
+    }
+
+    public int nodeId() {
+        return nodeId;
+    }
+
+    public String host() {
+        return host;
+    }
+
+    public int port() {
+        return port;
+    }
+
+    public VoterSet.VoterNode toVoterNode(String controllerListenerName) {
+        ReplicaKey voterKey = ReplicaKey.of(nodeId, directoryId);
+        Endpoints listeners = Endpoints.fromInetSocketAddresses(Collections.singletonMap(
+                new ListenerName(controllerListenerName),
+                new InetSocketAddress(host, port)));
+        SupportedVersionRange supportedKRaftVersion =
+                new SupportedVersionRange((short) 0, (short) 1);
```

Comment:
Okay. I am adding `SupportedVersionRange supportedVersionRange()` to `o.a.k.s.c.Features` in https://github.com/apache/kafka/pull/16735

Reply:
we can change it at that point then